### PR TITLE
irmin-pack: appending into the sparse file

### PR DESCRIPTION
**Pull request by art-w**
To allow updating the lower volumes :)

(This was a good opportunity to refactor the GC to use the new append functionality, and hence test that it works!)

### CODE REVIEW

File: src/irmin-pack/unix/sparse_file.ml [hunk 1]
```diff
@@ -324,15 +172,78 @@ module Make (Io : Io.S) = struct
...
+        let* end_poff = Io.size_of_path path in
+        Ao.open_rw ~path ~end_poff ~dead_header_size:0
+          ~auto_flush_threshold:1_000_000 ~auto_flush_procedure:`Internal
+      in
+      let* mapping = ao_open mapping in
+      let+ data = ao_open data in
+      { mapping; data }
+
+    let append_seq_exn t ~off seq =
+      (* TODO: check that [off] is larger or equal to the previous entry ending offset. *)
```
Comment (metanivek) [inline src/irmin-pack/unix/sparse_file.ml:224]: Reminder to capture this in the miscellaneous section of our überissue.

**Review by metanivek — COMMENTED** [review]
These changes look great! Lots of improvements. :tada: Left some small comments for consideration.

For the commit history, I think it would be nice to clean up some of the (slightly thrashy) changes of the append/write modifications in sparse file (looking at commit 2, 3, and 4). I could see either all of them being squashed into the final commit (essentially the new sparse file api + gc integration) or one commit before the GC integration that completely introduces the new API for sparse file (namely, `Wo` and `Ao` modules).

File: src/irmin-pack/unix/sparse_file.ml (hunk 1 above)
Comment (art-w) [reply src/irmin-pack/unix/sparse_file.ml]: I've added a sanity check (that should not raise if the future GC code is correct)

File: src/irmin-pack/unix/sparse_file.ml [hunk 2]
```diff
@@ -324,15 +172,78 @@ module Make (Io : Io.S) = struct
...
+          ~auto_flush_procedure:`Internal
+      in
+      let* mapping = ao_create mapping in
+      let+ data = ao_create data in
+      { mapping; data }
+
+    let open_ao ~mapping ~data =
+      let open Result_syntax in
+      let ao_open path =
+        (* TODO: the expected [end_poff] must be obtained from the control file *)
```
[the remark this answers is not in the document]
Comment (art-w) [reply src/irmin-pack/unix/sparse_file.ml]: Yes, done! It also checks that the data file has the right size (from the information available in the mapping file)

[the remark this answers is not in the document]
Comment (metanivek) [reply src/irmin-pack/unix/sparse_file.ml]: Ah, I see that yep! If we keep this pattern, do you think we should remove `data_poff` from the volume control file and just let `mapping_poff` drive it all?

**Review by metanivek — APPROVED** [review]

**Review by Firobe — APPROVED** [review]
That's great! The GC is simpler to grasp with the appropriate abstraction :+1:

File: src/irmin-pack/unix/sparse_file.ml (hunk 2 above)
[the remark this answers is not in the document]
Comment (art-w) [reply src/irmin-pack/unix/sparse_file.ml]: Yes the volume control file's `data_end_poff` is going to be redundant (and I think it's more important to check that the mapping has the right expectation). I'm not sure if this PR is the right moment to remove it, as it's going to trigger errors in other pending PRs?

**Pull request merged**